src/item_iterator.rs

```diff
@@ -154,52 +158,71 @@ fn items_in_container(item: &Item) -> Option<&Vec<Id>> {
-    PublicItem(PublicItemInner {
-        prefix: public_item.prefix(),
+    PublicItem {
         path: public_item
             .path()
             .iter()
             .map(|i| i.get_effective_name())
-            .collect::<Vec<String>>()
-            .join("::"),
-        suffix: public_item.suffix(),
-    })
+            .collect::<Vec<String>>(),
+        tokens: public_item.render_token_stream(),
+    }
 }
 
-/// To hide implementation details as much as possible from people who casually
-/// skims over the code in our lib.rs
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
-pub struct PublicItemInner {
-    /// The "pub struct/fn/..." part of an item.
-    pub(crate) prefix: String,
-
-    /// The "your_crate::mod_a::mod_b" part of an item.
-    pub(crate) path: String,
+/// Represent a public item of an analyzed crate, i.e. an item that forms part
+/// of the public API of a crate. Implements [`Display`] so it can be printed. It
+/// also implements [`Ord`], but how items are ordered are not stable yet, and
+/// will change in later versions.
+#[derive(Clone)]
+pub struct PublicItem {
+    /// The "your_crate::mod_a::mod_b" part of an item. Split by "::"
+    pub(crate) path: Vec<String>,
+
+    /// The rendered item into a stream of [`Token`]s
+    pub tokens: TokenStream,
+}
 
-    /// The type info part, e.g. "(param_a: Type, param_b: OtherType)" for a
-    /// `fn`.
-    pub(crate) suffix: String,
+/// We want pretty-printing (`"{:#?}"`) of [`crate::diff::PublicItemsDiff`] to print
+/// each public item as `Display`, so implement `Debug` with `Display`.
+impl std::fmt::Debug for PublicItem {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
 }
 
 /// One of the basic uses cases is printing a sorted `Vec` of `PublicItem`s. So
 /// we implement `Display` for it.
-impl Display for PublicItemInner {
+impl Display for PublicItem {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}{}{}", self.prefix, self.path, self.suffix)
+        write!(f, "{}", self.tokens)
     }
 }
 
-/// We want pretty-printing (`"{:#?}"`) of [`crate::diff::PublicItemsDiff`] to print
-/// each public item as `Display`, so implement `Debug` with `Display`.
-impl Debug for PublicItem {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        Display::fmt(self, f)
+impl PartialEq for PublicItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.path == other.path && self.tokens == other.tokens
+    }
+}
+
+impl Eq for PublicItem {}
+
+impl PartialOrd for PublicItem {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PublicItem {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self.path.cmp(&other.path) {
+            std::cmp::Ordering::Equal => self.tokens.cmp(&other.tokens),
+            ord => ord,
+        }
```

Comment:
I reverted the sort order, I liked the new one because it sorted related items (same path) together so the new enum and its fields would be close while in the previous one it sorted item kinds (fn/struct/etc) together. But this can be very easily adding in later.

For the two examples. the first one I deliberately deleted the `$` because that is not what a user would see in their own IDE and replaced it with a `Token` that is different from the standard, so it retains the information.

For the second one I will dive into how to restore that, good catch!